sklearn/metrics/cluster/adjacency_matrix.py

```diff
@@ -0,0 +1,22 @@
+import numpy as np
+
+
+def adjacency_matrix(cluster_assignement):
+    """
+    Parameter
+    ---------
+    cluster_assignement: vector (n_samples) of int i, 0 <= i < k
+
+    Return
+    ------
+    adj_matrix: matrix (n_samples, n_samples)
+        adji_matrix[i, j] = cluster_assignement[i] == cluster_assignement[j]
+    """
+    n_samples = len(cluster_assignement)
+    adj_matrix = np.zeros((n_samples, n_samples))
```

Comment:
It looks like this will require O(n^2) memory, O(n^2) time. This could be improved significantly if a sparse matrix were used.

Reply:
As long as we materialise the matrix in memory (even sparsely), this needs in worst-case `O(n^2)` time, but only because it needs to be `O(max_cluster_size^2 + n)`. Here's one sparse construction, assuming `y` is the cluster assignment and clusters are numbered from :
```python
# get an array of member samples for each cluster:
# [might be just as good to do this with a defaultdict(list)]
cluster_sizes = np.bincount(y)
cluster_members = np.split(np.argsort(y), np.cumsum(cluster_sizes))
lil = np.take(cluster_members, y)
indices = np.hstack(lil)
indptr = np.cumsum(np.hstack([0, np.take(cluster_sizes, y)]))
out = csr_matrix((np.ones_like(indptr), indices, indptr))
```

Another sparse construction with the same condition on y:
```
out = cosine_similarity(sp.csr_matrix((np.ones_like(y), y, np.arange(len(y)+1))), dense_output=False).astype(int)
```